### ISSUE TITLE
Flaky timeout behavior in spec

### ISSUE THREAD

**Issue by svroonland** [zio/zio#6231]
Using ZIO RC-1, the following test gets stuck, without failing due to the timeout of 30 seconds. 

Test is pretty much minified, the atomic integer is just to see visually that the test gets always stuck after the 1000 iterations, not somewhere halfway. Reducing the flaky number to say a 100 makes the test complete sometimes. 

Might be related to #6162, but I didn't want to add it to that issue because it might be a separate cause.

```scala
import zio.Console.printLine
import zio.test.Assertion._
import zio.test.TestAspect.nonFlaky
import zio.test._
import zio.{ZIO, _}

import java.util.concurrent.atomic.AtomicInteger

object TimeoutSpec extends DefaultRunnableSpec {
  val i = new AtomicInteger(0)

  override def spec = suite("TestClock")(
    test("has an issue somewhere") {
      val f = ZIO(i.addAndGet(1)).flatMap(nr => printLine(s"nr: ${nr}"))
      for {
        result <- f.timeoutFail("Timed out!")(10.seconds)
      } yield assertCompletes
    }
  ) @@ nonFlaky(1000)
}
```

**Comment by svroonland** [comment]
See also https://scastie.scala-lang.org/cJj6RsPES5GhGJhCnCvtLg

**Comment by adamgfraser** [comment]
Thanks for reporting! Will take a look!

**Comment by adamgfraser** [comment]
@svroonland This was caused by a regression that was fixed in #6192 and I confirmed locally that this test completes correctly on `series/2.x` so I think we can close this.

**Comment by svroonland** [comment]
Yes, I can verify that my original test code works with the latest snapshot. Thanks!
